### PR TITLE
Add Arm Neon intrinsics implementations for all SIMD paths

### PR DESCRIPTION
**Pull request by jonathanwright-ARM**
This set of patches won't build without an appropriate jconfigint.h file.

### CODE REVIEW

**Comment by ProgramMax** [comment]
These changes have been submitted to Chromium, where I have been doing code reviews for Jonathan: https://source.chromium.org/chromium/chromium/src/+/master:third_party/libjpeg_turbo/

I've been impressed with how much thought and care Jonathan as put into this work. He clearly knows a lot about libjpeg-turbo, jpegs in general, ARM optimization, and writing clear code.

I hope these changes are accepted here. The intrinsics allow compilers more room for broad optimization over assembly. And along the way, clean ups were made. This should be a universal win.

**Comment by widgetii** [comment]
Tested on my fleet of HiSilicon cameras, found a boost on modern devices (with NEON instructions):

| Chip                | Time  |
|---------------------|-------|
| 3516EV200 (no SIMD) | 1.05s |
| 3516EV200 (no SIMD) | 1.03s |
| 3516EV200 (-mcpu) `*` | 1.00s |
| 3516EV300 (-mcpu) `*` | 0.98s |
| 3516EV200 (this PR)  | 0.82s |
| 3516EV300 (this PR)  | 0.792s |

`*` - `-mcpu=cortex-a7 -mfloat-abi=softfp -mfpu=neon-vfpv4`

Great job!

**Comment by dcommander** [comment]
Unfortunately, the preliminary benchmark results on my Rock960 board are mixed:
https://www.dropbox.com/s/cp1dvdsqbx86dle/libjpegturbo-armintrinsics.ods?dl=0
(Note that I've personally tuned this platform to deliver accuracy within 1% for these benchmarks, so any differences greater than that are likely legitimate. I re-validate the accuracy each time I do a benchmarking run.)

Clang 10 performs well with 64-bit code (Clang apparently isn't available in an AArch32 version on Ubuntu/AArch64, so I can't test 32-bit code with that compiler), showing neutral or faster results across the board.  With GCC, however, the results are problematic.  32-bit performance mostly improves, but it has a handful of unacceptable regressions in decompression performance (anything > 5% is unacceptable), specifically when decompressing 4:2:2 or 4:4:4 images.  64-bit performance mostly regresses, and usually by a quite significant amount.  When we discussed this in the past, you indicated that GCC would not perform as well as Clang, but you also suggested that the speedup from accelerating additional SIMD algorithms with NEON would make up for any regression.  I'm not seeing that, unfortunately.  In your GCC 9.2 results, there are several cases in which adding an intrinsics implementation of a specific algorithm introduced a small slow-down, and those slow-downs compounded as intrinsics implementations of other algorithms were introduced.  What I would expect is that each algorithm implemented using intrinsics will produce a neutral or positive minimum, average, and maximum incremental speedup (that is, no algorithm in isolation will ever produce any significant slow-down.)  I held myself to that same standard when developing/integrating the AltiVec, AVX2, MMI, and 64-bit SSE2 extensions.

The amount of funding we agreed upon was predicated on the assumption that I wouldn't have to do anything other than a complete benchmarking run (mainly to understand the performance profile so that I can make reasonable claims about the speedup of the enhancement on various platforms), performing rudimentary code cleanup, and giving feedback.  Given how long it takes to do a complete benchmarking run, I don't have the budget to do any diagnosis or performance tuning.  At the moment, I unfortunately have to throw this back over the fence to you and see if there's any way to improve the performance with older compilers.  Given the far reach of libjpeg-turbo, we just can't assume that it will be built with the latest & greatest compilers.  ARM Linux distributions will generally build libjpeg-turbo with the default system compiler, and someone will definitely notice if the system's distribution of libjpeg-turbo regresses by 20-30%.  There are still actively supported distributions out there on which newer GCC or Clang releases aren't readily available.  Even Linaro's standalone toolchains are still on GCC 7.5.0.  Furthermore, even your GCC 9.2 results show a few unacceptable outliers.

Sorry, but this is a no go at the moment.

File: jmemmgr.c [hunk 1]
```diff
@@ -95,6 +95,15 @@ round_up_pow2(size_t a, size_t b)
...
 
+#ifdef WITH_SIMD
+#if (defined(__ARM_NEON) || defined(__ARM_NEON__)) && (ALIGN_SIZE % 32)
+  /* 32-byte alignment allows us to extract more performance from */
+  /* fancy-upsampling algorithms when using Neon. */
+  #error "Neon optimizations rely on 32-byte alignment. \
+          ALIGN_SIZE is not a multiple of 32 bytes."
+#endif
+#endif
+
```
Comment (dcommander) [inline jmemmgr.c:106]: @jonathanwright-ARM `ALIGN_SIZE` is always 32 if `WITH_SIMD` is defined, so I'm not sure why this code is necessary.  Please comment.

Comment (jonathanwright-ARM) [reply jmemmgr.c:106]: I read the comment starting on line 70 as saying that users can define ALIGN_SIZE themselves if they think their alignment needs are less than 32 bytes. This is technically true for, e.g. Arm systems with Neon enabled, and they may set ALIGN_SIZE = 16 thinking they'll save a few bytes here and there.
However, I have written the Neon optimizations for fancy-upsampling such that the assumed minimum output buffer size is 64 bytes (i.e. 2*ALIGN_SIZE). I added this extra compile-time check to make sure the alignment is what I think it is (the default here), else the code will likely scribble on memory it doesn't own.

Comment (dcommander) [reply jmemmgr.c:106]: That is true for the other SIMD extensions as well.  There is no way for anyone to change `ALIGN_SIZE` without modifying the libjpeg-turbo source code, and as soon as they do that, they're on their own as far as making it work.

Comment (jonathanwright-ARM) [reply jmemmgr.c:106]: It was extra caution on my part to make sure it doesn't fail silently if someone tries to be clever - if it's not warranted, discard the check.

Comment (dcommander) [reply jmemmgr.c:106]: Yeah, the ability to override build settings in **jconfig.h** is a legacy of the pre-autotools days when libjpeg was introduced.  No one realistically does that anymore.

**Comment by dcommander** [comment]
Pushed to dev, with numerous changes:
- Lots of wordsmithing (in the future, please don't over-describe the general JPEG algorithms.  They are described in detail elsewhere.  I expect a SIMD implementation of one of those algorithms to describe only how the specific implementation is unique.)
- Lots of formatting tweaks
- Lots of benchmarking, both to determine which GAS implementations were necessary to keep in order to retain performance parity with GCC relative to libjpeg-turbo 2.0.x, and to adjust the ultimate performance claims regarding the feature

Remaining issues:
- The full intrinsics implementation does not build with Visual Studio (it errors on **jcphuff-neon.c**, in part due to a missing `__builtin_clz()` function but also due to an inscrutable initialization error that I have been unable to diagnose.)
- I am having trouble determining exactly which versions of Clang support the necessary intrinsics.
    - With Xcode, 6.2.x is required (`Apple LLVM version 6.1.0 (clang-602.0.53) (based on LLVM 3.6.0svn)`.)  The [Xcode release notes](https://developer.apple.com/library/archive/documentation/Xcode/Conceptual/RN-Xcode-Archive/Chapters/xc6_release_notes.html) confirm that Xcode 6.2 introduced Apple LLVM v6.1 and changes to the Neon intrinsics, so that makes sense.
    - With the Android NDK, I can build the AArch64 code with r16b (either Clang, `Android clang version 5.0.300080  (based on LLVM 5.0.300080)`, or GCC 4.9.x), but I can't build the AArch32 code with Clang unless I use r19 or later (`Android (5058415 based on r339409) clang version 8.0.2 (https://android.googlesource.com/toolchain/clang 40173bab62ec746213857d083c0e8b0abb568790) (https://android.googlesource.com/toolchain/llvm 7a6618d69e7e8111e1d49dc9e7813767c5ca756a) (based on LLVM 8.0.2svn)`.)  Otherwise it complains about a missing `vld1_s16_x3()` intrinsic.
    - With Linux, I can build with Clang 3.9 or later (`clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)`.)
    - It would be nice to reconcile this somehow and perhaps add appropriate `#ifdef`s to ensure that the code at least builds on all of these configurations.  Perhaps the existing GCC `#ifdef`s can be repurposed, or perhaps the build system can test for the availability of particular intrinsics and react accordingly.  At this point, though, I just want to understand what's going on.

**Comment by dcommander** [comment]
The other issue I notice is that, regardless of whether the full Neon intrinsics implementation is used, `FLOATTEST=32bit` seems to be necessary with 64-bit Clang builds, but `FLOATTEST=64bit` seems to be necessary with 64-bit GCC builds.  Any explanation?  I would have thought that both types of builds are using Neon for floating point.  (?)

**Comment by dcommander** [comment]
I resolved the compatibility issues on my own, so the only remaining issues are `FLOATTEST` and the Windows build failure.

**Comment by jonathanwright-ARM** [comment]
> The other issue I notice is that, regardless of whether the full Neon intrinsics implementation is used, `FLOATTEST=32bit` seems to be necessary with 64-bit Clang builds, but `FLOATTEST=64bit` seems to be necessary with 64-bit GCC builds. Any explanation? I would have thought that both types of builds are using Neon for floating point. (?)

This issue has existed in libjpeg-turbo since I started working on it; it was not introduced by our contribution of intrinsics code.

IEEE 754 32-bit float should always produce the same results regardless of whether the system supports 32-bit and 64-bit so having a float test for 32-bit and 64-bit is not necessary given we're talking about cases where FAST_FLOAT is always defined to be "float" and not "double". 

The discrepancy between the 64-bit builds with GCC and Clang is due to GCC being overly aggressive about fusing multiply and add instructions into a single FMA instruction - which has different rounding compared to executing the two instructions sequentially. To turn this behaviour off, add "-ffp-contract=off" to the CFLAGS for 64-bit GCC builds - the 64-bit and 32-bit builds (for both Clang and GCC) will now produce the same results.

(I'm working on the MSVC issues.)

**Comment by jonathanwright-ARM** [comment]
On the MSVC issues:

- MSVC has its own version of `__builtin_clz `, defined as `_CountLeadingZeros`[1]. We'll need to `#ifdef` on `_MSV_VER` wherever we use `__builtin_clz`. We also have the option of using the macros defined here[2].
- It turns out that MSVC doesn't support static vector initialisation, i.e. `const uint16x8_t bias = { 0, 1, 0, 1, 0, 1, 0, 1 };`. This is the cause of the `error C2078: too many initializers` in jcsample-neon.c and jcphuff-neon.c. We'll need to define these vectors as const static buffers outside of their respective functions and explicitly load the values using `vld1q_<type>`.
- There may also be issues with supporting inline assembly for ARM64[3] so we may need a slow workaround for cases like this[4]. 

If you're happy for me to proceed, I'll put a few patches together implementing these changes and open a PR.

[1] https://docs.microsoft.com/en-us/cpp/intrinsics/arm64-intrinsics?view=msvc-160
[2] https://github.com/libjpeg-turbo/libjpeg-turbo/blob/f7ca3c5a3d4b8d2f3df4c7d3500901a9dafcb326/jchuff.c#L32
[3] https://docs.microsoft.com/en-us/cpp/assembler/inline/inline-assembler?view=msvc-160
[4] https://github.com/libjpeg-turbo/libjpeg-turbo/blob/dev/simd/arm/jchuff.h#L70

**Comment by dcommander** [comment]
> * MSVC has its own version of `__builtin_clz `, defined as `_CountLeadingZeros`[1]. We'll need to `#ifdef` on `_MSV_VER` wherever we use `__builtin_clz`. We also have the option of using the macros defined here[2].

I think it best to use MSVC's built-in equivalents and `#ifdef` around them.  That's what we do in the C version of progressive Huffman encoding:
https://github.com/libjpeg-turbo/libjpeg-turbo/blob/master/CMakeLists.txt#L358-L364
https://github.com/libjpeg-turbo/libjpeg-turbo/blob/master/jconfigint.h.in#L28-L34
https://github.com/libjpeg-turbo/libjpeg-turbo/blob/master/jcphuff.c#L26-L36
https://github.com/libjpeg-turbo/libjpeg-turbo/blob/master/jcphuff.c#L173-L182

> * It turns out that MSVC doesn't support static vector initialisation, i.e. `const uint16x8_t bias = { 0, 1, 0, 1, 0, 1, 0, 1 };`. This is the cause of the `error C2078: too many initializers` in jcsample-neon.c and jcphuff-neon.c. We'll need to define these vectors as const static buffers outside of their respective functions and explicitly load the values using `vld1q_<type>`.

Something like this seems to work:
```
#ifdef _MSC_VER
  uint8x8_t bitmap_mask;
  bitmap_mask.n64_i8[0] = 0x01;  bitmap_mask.n64_i8[1] = 0x02;
  bitmap_mask.n64_i8[2] = 0x04;  bitmap_mask.n64_i8[3] = 0x08;
  bitmap_mask.n64_i8[4] = 0x10;  bitmap_mask.n64_i8[5] = 0x20;
  bitmap_mask.n64_i8[6] = 0x40;  bitmap_mask.n64_i8[7] = 0x80;
#else
 const uint8x8_t bitmap_mask =
    { 0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80 };
#endif
```

**Comment by dcommander** [comment]
Run-time testing this is, however, another matter.  I personally have no way of running Windows/Arm.

**Comment by jonathanwright-ARM** [comment]
> > * MSVC has its own version of `__builtin_clz `, defined as `_CountLeadingZeros`[1]. We'll need to `#ifdef` on `_MSV_VER` wherever we use `__builtin_clz`. We also have the option of using the macros defined here[2].
> 
> I think it best to use MSVC's built-in equivalents and `#ifdef` around them. That's what we do in the C version of progressive Huffman encoding

Doing this for each instance of this builtin makes the code look quite untidy and cluttered. Would you rather me create a function macro `BUILTIN_CLZ()` and do the `#ifdef` just the once in its definition? If so, where ought this new macro be defined?

**Comment by dcommander** [comment]
neon-compat.h.in seems like a logical place.  Do you need to do likewise for `__builtin_clzl()`?  Does Arm have the capability to run-time test this code on an actually Windows/Arm device?

**Comment by jonathanwright-ARM** [comment]
> neon-compat.h.in seems like a logical place. 

Ok, sounds good.

> Do you need to do likewise for `__builtin_clzl()`? 

Yes.

> Does Arm have the capability to run-time test this code on an actual Windows/Arm device?

Yes.

**Comment by jonathanwright-ARM** [comment]
Right, I have a stack of patches that enable successful builds for Windows on Arm using MSVC.

However, for reasons that I am yet to determine, some tests fail:
`1>The following tests FAILED:
1>	  1 - tjunittest-shared (Failed)
1>	  2 - tjunittest-shared-alloc (Failed)
1>	  3 - tjunittest-shared-yuv (SEGFAULT)
1>	  4 - tjunittest-shared-yuv-alloc (SEGFAULT)
1>	  5 - tjunittest-shared-yuv-nopad (SEGFAULT)
1>	  9 - tjbench-shared-tile-gray-8x8-cmp (Failed)
1>	 13 - tjbench-shared-tile-gray-16x16-cmp (Failed)
1>	 17 - tjbench-shared-tile-gray-32x32-cmp (Failed)
1>	 21 - tjbench-shared-tile-gray-64x64-cmp (Failed)
1>	 25 - tjbench-shared-tile-gray-128x128-cmp (Failed)
1>	 73 - cjpeg-shared-gray-islow-cmp (Failed)
1>	 75 - djpeg-shared-gray-islow-cmp (Failed)
1>	 77 - djpeg-shared-gray-islow-rgb-cmp (Failed)
1>	 79 - djpeg-shared-gray-islow-565-cmp (Failed)
1>	 81 - djpeg-shared-gray-islow-565D-cmp (Failed)
1>	 87 - djpeg-shared-3x2-float-prog-cmp (Failed)
1>	156 - tjunittest-static (Failed)
1>	157 - tjunittest-static-alloc (Failed)
1>	158 - tjunittest-static-yuv (SEGFAULT)
1>	159 - tjunittest-static-yuv-alloc (SEGFAULT)
1>	160 - tjunittest-static-yuv-nopad (SEGFAULT)
1>	164 - tjbench-static-tile-gray-8x8-cmp (Failed)
1>	168 - tjbench-static-tile-gray-16x16-cmp (Failed)
1>	172 - tjbench-static-tile-gray-32x32-cmp (Failed)
1>	176 - tjbench-static-tile-gray-64x64-cmp (Failed)
1>	180 - tjbench-static-tile-gray-128x128-cmp (Failed)
1>	228 - cjpeg-static-gray-islow-cmp (Failed)
1>	230 - djpeg-static-gray-islow-cmp (Failed)
1>	232 - djpeg-static-gray-islow-rgb-cmp (Failed)
1>	234 - djpeg-static-gray-islow-565-cmp (Failed)
1>	236 - djpeg-static-gray-islow-565D-cmp (Failed)
1>	242 - djpeg-static-3x2-float-prog-cmp (Failed)`

The COVID-19 work from home situation means that I don't have easy access to a Windows on Arm device for testing at the moment - a colleague did the initial test run for me - and it will be a few days before I get a device delivered to start debugging properly myself. In the meantime, I can open a pull request containing the build-fix patches; or I can wait to submit the build fixes and bug fixes in one go - let me know what you want me to do.

**Comment by dcommander** [comment]
Let's get it working before integrating.  First try setting `JSIMD_FORCENONE=1` in the environment and see if the failures go away.  That will tell us whether the issues are in the SIMD extensions.  Next, send me the output from running the failing tests with `ctest -V`.  That may give us a clue where to look.

**Comment by jonathanwright-ARM** [comment]
This PR contains build fixes and fixes for the test failures above: https://github.com/libjpeg-turbo/libjpeg-turbo/pull/475

**Comment by dcommander** [comment]
@jwright-arm I would like to remove the phrase "All Rights Reserved" from all copyright headers in libjpeg-turbo 3.2, because it's a legally obsolete phrase and is confusing with respect to open source software.  Do you foresee any objection from Arm?

**Comment by jwright-arm** [comment]
I'll need to get the opinion of our legal team before I can comment.

**Comment by jwright-arm** [comment]
Hi @dcommander, we're happy for the "All Rights Reserved" wording to be removed from copyright headers.

**Pull request closed without merging**